contracts/workflows/RoyaltyWorkflows.sol

```diff
@@ -0,0 +1,232 @@
+        __UUPSUpgradeable_init();
+    }
+
+    /// @dev Sets the NFT contract beacon address.
+    /// @param newNftContractBeacon The address of the new NFT contract beacon.
+    function setNftContractBeacon(address newNftContractBeacon) external restricted {
+        if (newNftContractBeacon == address(0)) revert Errors.RoyaltyWorkflows__ZeroAddressParam();
+        RoyaltyWorkflowsStorage storage $ = _getRoyaltyWorkflowsStorage();
+        $.nftContractBeacon = newNftContractBeacon;
+    }
+
+    /// @notice Transfers royalties from royalty policy to the ancestor IP's royalty vault, takes a snapshot,
+    /// and claims revenue on that snapshot for each specified currency token.
+    /// @param ancestorIpId The address of the ancestor IP.
+    /// @param currencyTokens The addresses of the currency (revenue) tokens to claim (each address must be unique).
+    /// @param royaltyClaimDetails The details of the royalty claim from child IPs,
+    /// see {IRoyaltyWorkflows-RoyaltyClaimDetails}.
+    /// @return snapshotId The ID of the snapshot taken.
+    /// @return amountsClaimed The amount of revenue claimed for each currency token.
+    function transferToVaultAndSnapshotAndClaimByTokenBatch(
+        address ancestorIpId,
+        address[] calldata currencyTokens,
+        RoyaltyClaimDetails[] calldata royaltyClaimDetails
+    ) external returns (uint256 snapshotId, uint256[] memory amountsClaimed) {
+        (snapshotId, amountsClaimed, ) = _transferToVaultAndSnapshotAndClaimByTokenBatch(
+            ancestorIpId,
+            currencyTokens,
+            royaltyClaimDetails
+        );
+    }
+
+    /// @notice Transfers royalties to the ancestor IP's royalty vault, takes a snapshot, claims revenue for each
+    /// specified currency token both on the new snapshot and on each specified unclaimed snapshots.
+    /// @param ancestorIpId The address of the ancestor IP.
+    /// @param currencyTokens The addresses of the currency (revenue) tokens to claim (each address must be unique).
+    /// @param unclaimedSnapshotIds The IDs of unclaimed snapshots to include in the claim.
+    /// @param royaltyClaimDetails The details of the royalty claim from child IPs,
+    /// see {IRoyaltyWorkflows-RoyaltyClaimDetails}.
+    /// @return snapshotId The ID of the snapshot taken.
+    /// @return amountsClaimed The amounts of revenue claimed for each currency token.
+    function transferToVaultAndSnapshotAndClaimBySnapshotBatch(
+        address ancestorIpId,
+        address[] calldata currencyTokens,
+        uint256[] calldata unclaimedSnapshotIds,
+        RoyaltyClaimDetails[] calldata royaltyClaimDetails
+    ) external returns (uint256 snapshotId, uint256[] memory amountsClaimed) {
+        // Claims revenue for each specified currency token from the latest snapshot
+        IIpRoyaltyVault ancestorIpRoyaltyVault;
+        (snapshotId, amountsClaimed, ancestorIpRoyaltyVault) = _transferToVaultAndSnapshotAndClaimByTokenBatch(
```

Comment:
_transferToVaultAndSnapshotAndClaimByTokenBatch - is this function claiming by token batch inside an larger function which is by snapshot batch?

Perhaps for simplicity - given these internal functions are a bit too big - can we make the sequence of elementary functions transferToVault, snapshot, claimByTokenBatch and claimBySnapshotBatch without any internal functions first? Just to make sure we don't miss anything and it's easier to follow.

Then at the end we can think of making internal functions to reduce code duplication.

Reply:
For example the function `transferToVaultAndSnapshotAndClaimBySnapshotBatch` could be:

transferToVault()
snapshot()
claimBySnapshotBatch()

using more elementary building blocks. then at the end we think about internal functions.

Reply:
>_transferToVaultAndSnapshotAndClaimByTokenBatch - is this function claiming by token batch inside an larger function which is by snapshot batch?

Good question! All the SPG functions let you claim multiple tokens at the same time. For the ones using `claimBySnapshotBatch`, we first take the latest snapshot and claim all tokens, then handle any additional unclaimed snapshots user provide. We set it up this way so that if user don't need to bundle the calls, they can use the core protocol to claim your snapshots directly. 

One improvement we could add is making snapshotting optional in transferToVaultAndSnapshotAndClaimBySnapshotBatch and snapshotAndClaimBySnapshotBatch. That way, if taking the latest snapshot fails, users can still claim their unclaimed snapshots.

Reply:
> Perhaps for simplicity - given these internal functions are a bit too big - can we make the sequence of elementary functions transferToVault, snapshot, claimByTokenBatch / claimBySnapshotBatch without any internal functions first? Just to make sure we don't miss anything and it's easier to follow.

Agreed, that would make the code a lot more readable